File: python/delser.py
"""
http://www.brandonstaggs.com/2007/07/26/implementing-a-partial-serial-number-verification-system-in-delphi/
"""
import os
import json
import time
import warnings


class Delser(object):
    def __init__(self, byte_to_check=1, sequences=None, blacklist=None):

        if not os.path.exists(blacklist or 'blacklist.json'):
            warnings.warn("Could not locate blacklist")
            self.blacklist = set()
        else:
            with open(blacklist or 'blacklist.json', 'r') as fh:
                self.blacklist = [key.lower() for key in json.load(fh)]

        if not sequences:
            sequences = [
                (24, 3, 200),
                (10, 0, 56),
                (15, 2, 91),
                (25, 3, 200),
                (25, 3, 56)]

        self.sequences = sequences
        for cur_seq in self.sequences:
            assert len(cur_seq) == 3, 'Not enough values; "%s"' % cur_seq

        assert byte_to_check in range(len(self.sequences)), 'please choose a valid sequence to check against'
        self.byte_to_check = byte_to_check

        self._cached_checksums = {}
        self._cached_key_bytes = {}

    def _get_key_byte(self, seed: int, a: int, b: int, c: int):
        if (seed, a, b, c) not in self._cached_key_bytes:
            a = a % 25
            b = b % 3
            if a % 2 == 0:
                result = ((seed << a) & 0x000000FF) ^ ((seed << b) | c)
            else:
                result = ((seed << a) & 0x000000FF) ^ ((seed << b) & c)
            self._cached_key_bytes[(seed, a, b, c)] = result
        else:
            result = self._cached_key_bytes[(seed, a, b, c)]
        return result

    def get_checksum(self, string: str):

        if string not in self._cached_checksums:
            i = 0
            left = 0x0056
            right = 0x00AF
            if len(string) > 0:
                for i in range(len(string)):
                    right = right + ord(string[i])
                    if right > 0x00FF:
                        right -= 0x00FF
                    left += right
                    if left > 0x00FF:
                        left -= 0x00FF
            result = hex((left >> 8) + right)[2:].rjust(4, '0')
            self._cached_checksums[string] = result
        else:
            result = self._cached_checksums[string]
        return result

    def format_result(self, key: list):
        seed = key[0][2:].rjust(4, '0')
        key_bytes = '-'.join([x[2:].rjust(4, '0') for x in key[1:]])
        return seed + '-' + key_bytes

    def make_key(self, seed: int):
        key = []
        seed = int(seed)

        # the key string begins with a hexadecimal string of the seed
        key.append(hex(seed))

        # Fill keybytes with values derived from seed.
        # The parameters used here must be exactly the same
        # as the ones used in the check_key function.
        # A real key system should use more than four bytes.
        # then is followed by hexadecimal strings of each byte in the key
        for cur_seq in self.sequences:
            cur_byte = hex(self._get_key_byte(seed, *cur_seq) // 2)
            key.append(cur_byte)

        key = self.format_result(key)

        # add checksum to key string
        checksum = self.get_checksum(key)
        key += '-{}'.format(checksum)

        return key

    def check_key_checksum(self, key: str):

        if key.count('-') != (len(self.sequences) + 1):
            raise KeyInvalid('wrong number of sections; {} <> {}'.format(
                key.count('-') + 1, key))

        # remove cosmetic hyphens and normalize case
        key = key.upper().split('-')

        # last four characters are the checksum
        grabbed_checksum = key[-1]
        s = '-'.join(key[:-1])

        # compare the supplied checksum against the real checksum for
        # the key string.
        if not int(grabbed_checksum, 16) != int(self.get_checksum(s), 16):
            raise KeyBadChecksum('Incorrect checksum; {} is not equal to {}'.format(
                grabbed_checksum, self.get_checksum(s)))

        return True

    def check_key(self, key: str):
        # test against blacklist
        if len(self.blacklist) > 0:
            if key.lower() in self.blacklist:
                raise KeyBlacklisted(key)

        self.check_key_checksum(key)

        # remove cosmetic hyphens and normalize case
        key = key.upper().split('-')

        # At this point, the key is either valid or forged,
        # because a forged key can have a valid checksum.
        # We now test the "bytes" of the key to determine if it is
        # actually valid.

        # extract the seed from the supplied key string
        try:
            seed = int('0x' + key[0], 16)
        except ValueError:
            raise KeyPhony('Not hex; {}'.format(repr(key)))

        selected_sequence = self.sequences[self.byte_to_check]

        key_byte = key[self.byte_to_check + 1].upper()
        generated_byte = hex(self._get_key_byte(seed, *selected_sequence) // 2)[2:]

        if key_byte != generated_byte.upper().rjust(4, '0'):
            raise KeyPhony(key)

        # If we get this far, then it means the key is either good, or was made
        # with a keygen derived from "this" release.
        return True


class KeyGenException(Exception):
    pass


class KeyInvalid(KeyGenException):
    pass


class KeyBlacklisted(KeyGenException):
    pass


class KeyPhony(KeyGenException):
    pass


class KeyBadChecksum(KeyGenException):
    pass


def main():
    cur_delser = Delser()

    space = (65535, 524288)
    assert space[0] >= 0, 'the lower bound for the key must not be below zero'

    to_test = map(cur_delser.make_key, range(*space))

    bad = set()
    for test in to_test:
        if not cur_delser.check_key(test):
            print('For the key', test, 'with length', len(test))
            bad.add(test)
    print(len(bad), 'bad keys')


if __name__ == '__main__':
    # cur_delser = Delser()
    # while True:
    #     key = cur_delser.make_key(input('seed> '))
    #     cur_delser.check_key(key)
    #     print('key> {}\n'.format(key))
    main()

File: cpp/Delser.cpp
/*
http://www.brandonstaggs.com/2007/07/26/implementing-a-partial-serial-number-verification-system-in-delphi/
*/
#include <iostream>
#include <sstream>
#include <iomanip>

#include <memory>
#include <algorithm>

#include <vector>
#include <array>
#include <tuple>

#include "Delser.hpp"
#include "utils.cpp"

int Delser::_get_key_byte(int seed, int a, int b, int c) {
    a = a % 25;
    b = b % 3;
    long int result;
    if (a % 2 == 0) {
        result = ((seed << a) & 0x000000FF) ^ ((seed << b) | c);
    } else {
        result = ((seed << a) & 0x000000FF) ^ ((seed << b) & c);
    }
    return result;
}

std::string Delser::get_checksum(std::string string) {
    string = utils::upper(string);
    int left = 0x0056;
    int right = 0x00AF;
    if (string.length() > 0) {
        for (int i=0; i<string.length(); i++) {
            right = right + int(string[i]);
            if (right > 0x00FF) {
                right -= 0x00FF;
            }
            left += right;
            if (left > 0x00FF) {
                left -= 0x00FF;
            }
        }
    }
    std::string result = utils::int_to_hex((left >> 8) + right);
    return utils::upper(result);
}


std::string Delser::make_key(int seed) {
    std::vector<std::string> key;

    // the key string begins with a hexadecimal string of the seed
    key.push_back(utils::int_to_hex(seed));

    // Fill keybytes with values derived from seed.
    // The parameters used here must be exactly the same
    // as the ones used in the check_key function.
    // A real key system should use more than four bytes.
    // then is followed by hexadecimal strings of each byte in the key
    for (std::vector<sequence>::iterator cur_seq = sequences.begin(); cur_seq != sequences.end(); ++cur_seq) {
        int cur_key_byte = _get_key_byte(seed, std::get<0>(*cur_seq), std::get<1>(*cur_seq), std::get<2>(*cur_seq)) / 2;
        std::string cur_key_byte_str = utils::int_to_hex(cur_key_byte);
        key.push_back(cur_key_byte_str);
    }
    std::string skey = format_result(key);

    // add checksum to key string
    std::string checksum = get_checksum(skey);
    skey += "-" + checksum;

    return skey;
}

bool Delser::check_key_checksum(std::string skey) {
    if (std::count(skey.begin(), skey.end(), '-') != (sequences.size() + 1)) {
        throw key_invalid("wrong number of sections");//; %d <> %d", 
//            std::count(skey.begin(), skey.end(), '-') + 1, skey);
    }

    // remove cosmetic hyphens and normalize case
    std::vector<std::string> key = utils::split(utils::upper(skey), '-');
    
    // last four characters are the checksum
    std::string grabbed_checksum = key[key.size() - 1];
    
    // grab checksum
    std::vector<std::string> key_minus_checksum(key.begin(), key.end() - 1);

    // join vector with '-'
    std::string s = utils::vector_to_string(key_minus_checksum, '-');
    
    // compare the supplied checksum against the real checksum for
    // the key string.
    if (grabbed_checksum != get_checksum(s)) {
        throw key_bad_checksum("Incorrect checksum");//; %s is not equal to %s", 
            //grabbed_checksum, get_checksum(s));
    }

    return true;
}

bool Delser::check_key(std::string key) {
    // test against blacklist
    if (!blacklist.empty()) {
        if (std::find(blacklist.begin(), blacklist.end(), utils::upper(key)) != blacklist.end()) {
            throw key_blacklisted(key);
        }
    }

    check_key_checksum(key);

    // remove cosmetic hyphens and normalize case
    auto skey = utils::split(utils::upper(key), '-');

    // At this point, the key is either valid or forged,
    // because a forged key can have a valid checksum.
    // We now test the "bytes" of the key to determine if it is
    // actually valid.

    // extract the seed from the supplied key string
    int seed = utils::hex_to_int(skey[0]);

//    throw Exception()
    sequence selected_sequence = sequences[byte_to_check];

    std::string key_byte = utils::upper(skey[byte_to_check + 1]);
    std::string generated_byte = utils::int_to_hex(_get_key_byte(seed,
       std::get<0>(selected_sequence),
       std::get<1>(selected_sequence),
       std::get<2>(selected_sequence)) / 2);
    
    if (utils::rjust(key_byte, '0', generated_byte.length()) != utils::rjust(utils::upper(generated_byte), '0', 4)) {
        throw key_phony("Phony key");//skey);
    }

    // If we get this far, then it means the key is either good, or was made
    // with a keygen derived from "this" release.
    return true;
}

std::string Delser::format_result(std::vector<std::string>key) {
    std::string seed = utils::rjust(key[0], '0', 4);
    std::string key_bytes;

    for (int x=1; x<key.size(); x++) {
        key_bytes += '-' + utils::rjust(key[x], '0', 4);
    }
    
    return seed + key_bytes;
}

